src/builder.rs

```diff
@@ -0,0 +1,54 @@
+use crate::{errors::Error, Layer};
+
+#[derive(Debug)]
+pub struct PacketBuilder {
+    layers: Vec<Box<dyn Layer + Send>>,
+    raw_bytes: Option<Vec<u8>>,
+}
+
+impl PacketBuilder {
+    pub fn new() -> Self {
+        PacketBuilder {
+            layers: vec![],
+            raw_bytes: None,
+        }
+    }
+
+    pub fn stack(mut self, layer: Box<dyn Layer + Send>) -> Result<Self, Error> {
```

Comment:
This is a good API, One way to `restrict` complicated handling is - you can `stack` layers up. What that means? you can only do `PacketBuilder::stack(eth).stack(ip)` and so on. This would mean `stack`ing incorrect layers is an error. Also handling this is a bit interesting. Let's take an example - 

Let's say you do `stack(eth).stack(ip)` -> What this should do is pop the last layer , update it's `demux` and push that layer and the current layer. Concretely - each `layer` should implement a `stack` function that takes `mut self` and an `&str` and returns a `Self`. The `&str` (through an internal map) would update the `Ethertype` or `protocol` or whatever. 

This allows one to build packet layer by layer. I think if we implement this `stack` properly. then `build` can simply output the `Packet` structure that is *kind of* finalized. With addition of each new *layer* the internal metadata can be updated. 

Does this make sense?